### PR TITLE
 [TIMOB-26924] Support Visual Studio 2019

### PR DESCRIPTION
**Pull request by infosia**
- [TIMOB-26924](https://jira.appcelerator.org/browse/TIMOB-26924) Support Visual Studio 2019
- [TIMOB-26588](https://jira.appcelerator.org/browse/TIMOB-26588) Support Windows 10 SDK 10.0.17763.0
- [TIMOB-26019](https://jira.appcelerator.org/browse/TIMOB-26019) Support x64 build target
- [TIMOB-25999](https://jira.appcelerator.org/browse/TIMOB-25999) Support x64 build target for modules

Add support for Visual Studio 2019. This also enables `x64`and `ARM` architecture build target using `--architecture` option.

### Prerequisites

- Visual Studio Community 2019 GA
- Visual Studio Community 2017 version 15.9.11 Servicing Update to resolve [TIMOB-26588](https://jira.appcelerator.org/browse/TIMOB-26588) 

NOTE: Installing Visual Studio 2019 may cause build errors on existing Titanium SDKs (without applying this PR) because VS 2019 updates some build tools that do have breaking changes. You might want to prepare fresh Windows installation to test for QE this because VS 2019 installation will break builds on existing Titanium SDKs.

### Build with Visual Studio 2019

```
appc ti clean
appc run -p windows --target ws-local -l trace -V "Visual Studio Community 2019" --forceUnInstall
```

Expected: Build goes well and application should launch.

### Build with Visual Studio 2017

```
appc ti clean
appc run -p windows --target ws-local -l trace -V "Visual Studio Community 2017" --forceUnInstall
```

Expected: Build goes well and application should launch.

### Build with Visual Studio 2015

```
appc ti clean
appc run -p windows --target ws-local -l trace -V 14.0 --forceUnInstall
```

Expected: Build goes well and application should launch.

### Build with x64 bit ([TIMOB-26019](https://jira.appcelerator.org/browse/TIMOB-26019))

Note that you may need to uninstall previously-installed application manually before the build with x64 if you have installed it with different architecture previously.

```
appc ti clean
appc run -p windows --target ws-local -l trace -V "Visual Studio Community 2019" --forceUnInstall --architecture x64
```

Expected: Build goes well and application should launch.

### Build with Windows SDK 10.0.17763.0 ([TIMOB-26588](https://jira.appcelerator.org/browse/TIMOB-26588))

This actually just requires to update to latest version of Visual Studio. See Prerequisites section for details.

#### tiapp.xml
```xml
<windows>
    <TargetPlatformVersion>10.0.17763.0</TargetPlatformVersion>
    <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
</windows>
```

```
appc ti clean
appc run -p windows --target ws-local -l trace -V "Visual Studio Community 2017" --forceUnInstall
```

Expected: Build goes well and application should launch.

### x64 module ([TIMOB-25999](https://jira.appcelerator.org/browse/TIMOB-25999))

1. Create new module with `appc new`.
2. Build the module (`cd MODULE_PROJECT; cd windows;  appc ti build -p windows --build-only -l trace`
3. Use the module in `x64` project

### CODE REVIEW

**Comment by garymathews** [comment]
Could we get VS2019 on Jenkins and update our [Jenkins](https://github.com/appcelerator/titanium_mobile_windows/blob/master/Jenkinsfile#L188) file?

**Comment by infosia** [comment]
Makes sense. It should done after [TIMOB-26924](https://jira.appcelerator.org/browse/TIMOB-26924) is resolved. Created [TIMOB-26995](https://jira.appcelerator.org/browse/TIMOB-26995) to track this.

File: Tools/Scripts/generate_project.js [hunk 1]
```diff
@@ -44,18 +42,17 @@ if (os.platform() === 'win32') {
 function generateProject(example_name, dest, platform, sdkVersion, msdev, arch) {
 	return new Promise((resolve, reject) => {
 		// cmake generator name
-		let generator = (msdev === MSBUILD_14) ? VS_2015_GENERATOR : VS_2017_GENERATOR;
-		if (arch === 'ARM') {
-			generator += ' ARM';
-		}
+		let generator = VS_2017_GENERATOR;
```
Comment (garymathews) [inline Tools/Scripts/generate_project.js]: 2019 generator?

File: templates/module/default/hooks/windows-module.js [hunk 2]
```diff
@@ -130,14 +127,23 @@ function runCMake(logger, cmake, projectDir, targetEnv, targetArch, targetGenera
...
-		],
+		];
+
+	if (targetArch === 'ARM') {
+		cmakeArgs.push('-A', 'ARM');
+	} else if (targetArch === 'ARM64') {
+		cmakeArgs.push('-A', 'ARM64');
+	} else if (targetArch === 'x64') {
+		cmakeArgs.push('-A', 'x64');
+	}
```
Comment (ewanharris) [inline templates/module/default/hooks/windows-module.js]: When generating the project I'm seeing the Win32 project get generated as a x64 project which then causes the build of the project to fail. Adding the below causes this the correct project type to be generated 
```suggestion
	} else if (targetArch === 'Win32') {
		cmakeArgs.push('-A', 'Win32');
	}
```

Comment (infosia) [reply templates/module/default/hooks/windows-module.js]: Nice catch 👍 Fixed.

**Comment by infosia** [comment]
> FR Failed due to the Win32 module build issue I commented on, but other than that looking all good!
> 
> Not a blocker but I'm seeing the below warning when building to ws-local, I'm guessing something changed in the certificate within cmake that we use? Would be good to fix in a follow up release maybe
> 
> ```
> [WARN]  Build succeeded.
> 
> "C:\Users\awam\.titanium\vsbuild\testnewvs\win10.x86\Testnewvs.sln" (default target) (1) ->
> "C:\Users\awam\.titanium\vsbuild\testnewvs\win10.x86\Testnewvs.vcxproj.metaproj" (default target) (2) ->
> "C:\Users\awam\.titanium\vsbuild\testnewvs\win10.x86\Testnewvs.vcxproj" (default target) (6) ->
> (_GenerateCurrentProjectAppxManifest target) ->
>   C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(2649,5): warning : Publisher name (CN=00000000-0000-1000-8000-000000000000) does not match signing certificate subject: CN=CMake Test Cert. Updating Publisher name. [C:\Users\awam\.titanium\vsbuild\testnewvs\win10.x86\Testnewvs.vcxproj]
> 
>     1 Warning(s)
>     0 Error(s)
> 
> Time Elapsed 00:00:22.88
> ```

I was able to reproduce this. It seems latest Visual Studio 2019 starts complaining about the publisher name, but the problem is we are using test cert for test deployment and the cert actually not quite follows the required format for the Store publisher name (`CN=Test Cert`). It seems Visual Studio fixed the issue automatically for now so we can proceed this. Yes it would be good to have better cert for the test deployment in the followup release.

**Pull request merged**